File: app/kb_client.py
import json
from pathlib import Path
from typing import Dict, List, Optional
from difflib import SequenceMatcher


class KBClient:
    def __init__(self, data_path: Optional[Path] = None):
        base = Path(__file__).resolve().parent
        self.data_path = data_path or (base / "data" / "responses.json")
        self.entries: List[Dict] = []
        self._load()

    def _load(self):
        try:
            with open(self.data_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                self.entries = data.get('entries', [])
        except Exception:
            self.entries = []

        # Precompute a searchable text for each entry
        for entry in self.entries:
            kws = entry.get('keywords', []) or []
            combined = ' '.join(kws) + ' ' + entry.get('response', '')
            entry['_search_text'] = combined.lower()

    def reload(self):
        """Reload the dataset from disk."""
        self._load()

    def query(self, text: str) -> Dict:
        """Return the best matching entry for the text using fuzzy matching.

        Scoring combines keyword overlap and sequence similarity. Returns the best
        entry or a default fallback if nothing matches well.
        """
        text_lower = text.lower()

        best = None
        best_score = 0.0

        for entry in self.entries:
            # keyword overlap score
            keywords = entry.get('keywords', []) or []
            kw_score = sum(1 for kw in keywords if kw in text_lower)

            # sequence similarity against precomputed text
            search_text = entry.get('_search_text', '')
            seq = 0.0
            if search_text:
                seq = SequenceMatcher(None, text_lower, search_text).ratio()

                # combine scores: weight both keyword overlap and sequence similarity
                score = kw_score * 2.0 + seq * 2.0

            if score > best_score:
                best_score = score
                best = entry

        # Always return the best match if one exists; prefer non-empty keyword entries
        if best:
            return best

        # fallback to default entry (empty keywords) or a simple reply
        for entry in self.entries:
            if not entry.get('keywords'):
                return entry

        return {"response": "I'm here to listen.", "risk_level": "low"}

File: backend/src/services/KBClient.js
const fs = require('fs');
const path = require('path');
const stringSimilarity = require('string-similarity'); // Will need to npm install string-similarity

class KBClient {
    constructor(dataPath = null) {
        this.dataPath = dataPath || path.join(__dirname, '../../data', 'responses.json');
        this.entries = [];
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(this.dataPath)) {
                const data = JSON.parse(fs.readFileSync(this.dataPath, 'utf8'));
                this.entries = data.entries || [];
            } else {
                console.warn("responses.json not found:", this.dataPath);
                this.entries = [];
            }
        } catch (error) {
            console.error("Error loading responses:", error);
            this.entries = [];
        }

        // Precompute searchable text
        this.entries.forEach(entry => {
            const kws = entry.keywords || [];
            const combined = `${kws.join(' ')} ${entry.response || ''}`;
            entry._search_text = combined.toLowerCase();
        });
    }

    reload() {
        this.load();
    }

    query(text) {
        const textLower = text.toLowerCase();
        let best = null;
        let bestScore = 0;

        for (const entry of this.entries) {
            const keywords = entry.keywords || [];
            const kwScore = keywords.reduce((score, kw) => score + (textLower.includes(kw) ? 1 : 0), 0);
            
            let seqScore = 0;
            if (entry._search_text) {
                // simple substring/likeness similarity using string-similarity package
                seqScore = stringSimilarity.compareTwoStrings(textLower, entry._search_text);
            }

            const score = (kwScore * 2.0) + (seqScore * 2.0);

            if (score > bestScore) {
                bestScore = score;
                best = entry;
            }
        }

        if (bestScore > 0 && best) return best;

        // fallback to default entry (empty keywords)
        for (const entry of this.entries) {
            if (!entry.keywords || entry.keywords.length === 0) {
                return entry;
            }
        }

        return { response: "I'm here to listen.", risk_level: "low" };
    }
}

module.exports = KBClient;
